v3/src/models/data/plotted-value-formula-adapter.ts

```diff
@@ -0,0 +1,146 @@
+      if (mstAction.name === "setAttributeID") {
+        recalculate()
+      }
+    })
+    return dispose
+  }
+
+  getAllFormulas(): ({ formula: IFormula, extraMetadata?: IPlottedValueFormulaExtraMetadata })[] {
+    const result: ({ formula: IFormula, extraMetadata: IPlottedValueFormulaExtraMetadata })[] = []
+    this.graphContentModels.forEach(graphContentModel => {
+      graphContentModel.adornments.forEach(adornment => {
+        if (graphContentModel.dataset && isPlottedValueAdornment(adornment)) {
+          result.push({
+            formula: adornment.formula,
+            extraMetadata: {
+              graphContentModelId: graphContentModel.id,
+              dataSetId: graphContentModel.dataset.id,
+            }
+          })
+        }
+      })
+    })
+    return result
+  }
+
+  recalculateFormula(formulaContext: IFormulaContext, extraMetadata: IPlottedValueFormulaExtraMetadata) {
+    const { graphContentModelId } = extraMetadata
+    const graphContentModel = this.graphContentModels.get(graphContentModelId)
+    if (!graphContentModel) {
+      throw new Error(`GraphContentModel with id "${graphContentModelId}" not found`)
+    }
+    const adornment = graphContentModel.adornments.find(a => a.type === ADORNMENT_TYPE)
+    if (!adornment || !isPlottedValueAdornment(adornment)) {
+      throw new Error(`Adornment of type "${ADORNMENT_TYPE}" not found`)
+    }
+    // This code is mostly copied from UnivariateMeasureAdornmentModel.updateCategories.
+    // TODO: Is there a way to share it somehow?
+    const options = graphContentModel.getUpdateCategoriesOptions()
+    const { xCats, yCats, topCats, rightCats, resetPoints, dataConfig } = options
+    if (!dataConfig) return
+    const topCatCount = topCats.length || 1
+    const rightCatCount = rightCats.length || 1
+    const xCatCount = xCats.length || 1
+    const yCatCount = yCats.length || 1
+    const columnCount = topCatCount * xCatCount
+    const rowCount = rightCatCount * yCatCount
+    const totalCount = rowCount * columnCount
+    // TODO: use it as a default argument while working on this feature?
+    // const attrId = dataConfig.primaryAttributeID
+    for (let i = 0; i < totalCount; ++i) {
+      const cellKey = adornment.cellKey(options, i)
+      const instanceKey = adornment.instanceKey(cellKey)
+      const cases = dataConfig.subPlotCases(cellKey)
+      const value = Number(this.computeFormula(formulaContext, extraMetadata, cases))
+      if (!adornment.measures.get(instanceKey) || resetPoints) {
+        adornment.addMeasure(value, instanceKey)
+      } else {
+        adornment.updateMeasureValue(value, instanceKey)
+      }
+    }
+  }
+
+  computeFormula(formulaContext: IFormulaContext, extraMetadata: IPlottedValueFormulaExtraMetadata,
+    childMostCases: ICase[]) {
+    const { formula, dataSet } = formulaContext
+    console.log(`[plotted value formula] recalculate "${formula.canonical}"`)
```

Comment:
>> I know you said this is a first iteration but just in case: there are console.log statements in attribute-formula-adapter.ts and plotted-value-formula-adapter.ts that don't seem necessary. But maybe you want to keep them for now?
>
>These logs indicate when formulas are recalculated. If they are not too annoying for others, I'd like to keep them for now because it's probably the only way I can easily spot when a formula is recalculated when it shouldn't be. I remember @kswenson suggested using a debug toggle to show/hide them, but I'm worried that, in practice, no one will notice unwanted recalculations if we do that. So, I'll keep them in the near future if that's okay with everyone.

My inclination would be to wrap these in `if (DEBUG_FORMULAS)` and then if you want to have these logged by default for the time being we can default `DEBUG_FORMULAS` to `true` in `debug.ts`, but at least then there's only one place to change it rather than multiple places.

Reply:
Okay, I've updated debug.ts and followed the existing pattern. So, console logs won't be visible by default now. I'll keep that debug option in my local storage.